airflow/providers_manager.py

```diff
@@ -46,42 +50,136 @@ class ProvidersManager:
+            self._validator.validate(provider_info)
+            provider_info_package_name = provider_info['package-name']
+            if package_name != provider_info_package_name:
+                raise Exception(
+                    f"The package '{package_name}' from setuptools and "
+                    f"{provider_info_package_name} do not match. Please make sure they are"
+                    f"aligned"
+                )
+            if package_name not in self._provider_set:
+                self._provider_set.add(package_name)
+                self._provider_list.append((package_name, version, provider_info))
+            else:
+                log.warning(
+                    "The providers for package '%s' could not be registered because providers for that "
+                    "package name have already been registered",
+                    package_name,
+                )
+
+    def __find_all_airflow_builtin_providers_from_local_sources(self) -> None:
+        """
+        Finds all built-in airflow providers if airflow is run from the local sources.
+        It finds `provider.yaml` files for all such providers and registers the providers using those.
+
+        This 'provider.yaml' scanning takes precedence over scanning packages installed because
+        in case you have both sources and packages installed, the providers will be loaded from
+        the "airflow" sources rather than from the packages.
+        """
+        import airflow.providers
+
+        try:
+            for path in airflow.providers.__path__:
+                self.__add_provider_info_from_local_source_files_on_path(path)
+        except Exception as e:  # noqa pylint: disable=broad-except
+            log.warning("Error when loading 'provider.yaml' files from airflow sources: %s", e)
+
+    def __add_provider_info_from_local_source_files_on_path(self, path) -> None:
+        """
+        Finds all the provider.yaml files in the directory specified.
+        :param path: path where to look for provider.yaml files
+        """
+        root_path = path
+        for folder, _, files in os.walk(path):
+            for filename in fnmatch.filter(files, "provider.yaml"):
+                self.__add_provider_info_from_local_source_file(filename, folder, root_path)
+
+    def __add_provider_info_from_local_source_file(self, filename, folder, root_path) -> None:
+        """
+        Parses found provider.yaml file and adds found provider to the list of all providers.
+        :param filename: name of the provider.yaml file
+        :param folder: folder where the provider.yaml file is
+        :param root_path: root path where we started to search ("airflow/providers" folder from the
+                          local sources
+        """
+        try:
+            with open(os.path.join(folder, filename)) as provider_yaml_file:
+                provider_info = yaml.safe_load(provider_yaml_file.read())
```

Comment:
Can pass the stream directly to pyyaml:

```suggestion
                provider_info = yaml.safe_load(provider_yaml_file)
```

Reply:
True